src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -2027,45 +2121,47 @@ private void clientHandshakeSSO(final String db, final OpenIDAuthenticator authe
 			}
 
 			// Now send the tokens to the database for validation
-			clientHandshakeGCM(username, password, db, shouldRequestVersion, HandshakeType.SSO);
+			SSOHandshakeCompletionStage = clientHandshakeGCM(username, password, db, shouldRequestVersion, HandshakeType.SSO)
+											.thenApply(successOrEx -> {
+												LOGGER.log(Level.INFO, () -> "Handshake SSO without token finished");
+												return successOrEx;
+											});
 		}
 		catch(RuntimeException e)
 		{
-			LOGGER.log(Level.WARNING, String.format("Runtime Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
+			LOGGER.log(Level.WARNING, e, () -> String.format("Runtime Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
 			throw e;
 		}
 		catch (final Exception e)
 		{
-			LOGGER.log(Level.WARNING, String.format("Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
+			LOGGER.log(Level.WARNING, e, () -> String.format("Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
 
-			try
-			{
-				closeTransport();
-			}
-			catch (final Exception f)
-			{
-				LOGGER.log(Level.WARNING, "Failed to close socket in clientHandshakeSSO");
-			}
+			SSOHandshakeCompletionStage = SSOHandshakeCompletionStage
+				.thenCompose(handshakeSuccessOrEx -> handshakeSuccessOrEx
+				.recoverAsync(ex -> runAsyncSafe_FIXME(this::closeTransport)
+									.thenApply(closeTransportSuccessOrEx -> { 
 
-			throw e;
-		}
+										if(closeTransportSuccessOrEx.isException())
+											LOGGER.log(Level.WARNING, closeTransportSuccessOrEx.getException(), () -> "Failed to close socket in clientHandshakeSSO");
 
-		if (shouldRequestVersion)
-		{
-			if (serverVersion.equals(""))
-			{
-				// FIXME make async
-				blockingGetSQLResult(fetchServerVersion()).get();
-			}
+										return handshakeSuccessOrEx;
+									})));
 		}
-		LOGGER.log(Level.INFO, "Handshake SSO without token finished");
+
+		if (shouldRequestVersion && serverVersion.equals(""))
+			SSOHandshakeCompletionStage = SSOHandshakeCompletionStage.thenCompose(successOrEx -> successOrEx.mapSuccessAsync(success -> fetchServerVersion()));
+		
+		return SSOHandshakeCompletionStage;
 	}
 
 	public static OAuthFlow getDefaultOAuthFlow() {
 		return Desktop.isDesktopSupported() ? OAuthFlow.AUTHORIZATION_CODE : OAuthFlow.DEVICE_GRANT;
 	}
 
-	private void clientHandshakeOktaNativeSSO(final String db, final OpenIDAuthenticator authenticator, final boolean shouldRequestVersion) throws Exception {
+	private CompletionStage<SQLResult<Void>> clientHandshakeOktaNativeSSO(final String db, final OpenIDAuthenticator authenticator, final boolean shouldRequestVersion) throws Exception {
+
+		CompletionStage<SQLResult<Void>> OktaNativeSSOCompletionStage = SQLResult.voidFuture();
```

Comment:
(uber uber nit)

lower camel case for local variables. We should probably come up with a style guide for the library...
```suggestion
		CompletionStage<SQLResult<Void>> oktaNativeSSOCompletionStage = SQLResult.voidFuture();
```